# Practicing with pandas - 3
## Merging, Joining and Null Values

In [1]:
import pandas as pd
import numpy as np

In [2]:
#load the data into a panda dataframe
#Air quality dataset downloaded from http://archive.ics.uci.edu/ml/machine-learning-databases/00360/
df = pd.read_csv('AirQualityUCI.csv', sep=';', decimal=",")

In [3]:
#look at the first few rows to make sure the data was loaded correctly
df.head()

Date      Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  \
0  10/03/2004  18.00.00     2.6       1360.0     150.0      11.9   
1  10/03/2004  19.00.00     2.0       1292.0     112.0       9.4   
2  10/03/2004  20.00.00     2.2       1402.0      88.0       9.0   
3  10/03/2004  21.00.00     2.2       1376.0      80.0       9.2   
4  10/03/2004  22.00.00     1.6       1272.0      51.0       6.5   

   PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  \
0         1046.0    166.0        1056.0    113.0        1692.0       1268.0   
1          955.0    103.0        1174.0     92.0        1559.0        972.0   
2          939.0    131.0        1140.0    114.0        1555.0       1074.0   
3          948.0    172.0        1092.0    122.0        1584.0       1203.0   
4          836.0    131.0        1205.0    116.0        1490.0       1110.0   

      T    RH      AH  Unnamed: 15  Unnamed: 16  
0  13.6  48.9  0.7578          NaN          NaN  
1  13.3  47.7  0.7255          NaN          NaN  
2  11.9  54.0  0.7502          NaN          NaN  
3  11.0  60.0  0.7867          NaN          NaN  
4  11.2  59.6  0.7888          NaN          NaN

In [4]:
#Look at the last few rows
df.tail()

Date Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  PT08.S2(NMHC)  \
9466  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9467  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9468  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9469  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9470  NaN  NaN     NaN          NaN       NaN       NaN            NaN   

      NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)   T  RH  AH  \
9466      NaN           NaN      NaN           NaN          NaN NaN NaN NaN   
9467      NaN           NaN      NaN           NaN          NaN NaN NaN NaN   
9468      NaN           NaN      NaN           NaN          NaN NaN NaN NaN   
9469      NaN           NaN      NaN           NaN          NaN NaN NaN NaN   
9470      NaN           NaN      NaN           NaN          NaN NaN NaN NaN   

      Unnamed: 15  Unnamed: 16  
9466          NaN          NaN  
9467          NaN          NaN  
9468          NaN          NaN  
9469          NaN          NaN  
9470          NaN          NaN

In [5]:
#use info() function to find the number of missing values in each column
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9471 entries, 0 to 9470
Data columns (total 17 columns):
Date             9357 non-null object
Time             9357 non-null object
CO(GT)           9357 non-null float64
PT08.S1(CO)      9357 non-null float64
NMHC(GT)         9357 non-null float64
C6H6(GT)         9357 non-null float64
PT08.S2(NMHC)    9357 non-null float64
NOx(GT)          9357 non-null float64
PT08.S3(NOx)     9357 non-null float64
NO2(GT)          9357 non-null float64
PT08.S4(NO2)     9357 non-null float64
PT08.S5(O3)      9357 non-null float64
T                9357 non-null float64
RH               9357 non-null float64
AH               9357 non-null float64
Unnamed: 15      0 non-null float64
Unnamed: 16      0 non-null float64
dtypes: float64(15), object(2)
memory usage: 1.2+ MB


The data was loaded into a dataframe with 9471 observations of 17 variables. However, this dataset contains a large number of missing values. The last two unnamed columns do not have any non-null values. In addition, the dataframe contains several empty rows in the end. According to the metadata for this dataset, it should contain 9358 observations of 15 variables. These empty rows and columns are likely a result of a technical error introduced while merging sensor data in one data file. So, we can just delete the empty columns and rows.

In [6]:
#Drop the last two columns
df.drop(['Unnamed: 15', 'Unnamed: 16'], axis=1, inplace=True)

In [7]:
#Check results
df.head()

Date      Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  \
0  10/03/2004  18.00.00     2.6       1360.0     150.0      11.9   
1  10/03/2004  19.00.00     2.0       1292.0     112.0       9.4   
2  10/03/2004  20.00.00     2.2       1402.0      88.0       9.0   
3  10/03/2004  21.00.00     2.2       1376.0      80.0       9.2   
4  10/03/2004  22.00.00     1.6       1272.0      51.0       6.5   

   PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  \
0         1046.0    166.0        1056.0    113.0        1692.0       1268.0   
1          955.0    103.0        1174.0     92.0        1559.0        972.0   
2          939.0    131.0        1140.0    114.0        1555.0       1074.0   
3          948.0    172.0        1092.0    122.0        1584.0       1203.0   
4          836.0    131.0        1205.0    116.0        1490.0       1110.0   

      T    RH      AH  
0  13.6  48.9  0.7578  
1  13.3  47.7  0.7255  
2  11.9  54.0  0.7502  
3  11.0  60.0  0.7867  
4  11.2  59.6  0.7888

In [8]:
#look for the missing values again
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9471 entries, 0 to 9470
Data columns (total 15 columns):
Date             9357 non-null object
Time             9357 non-null object
CO(GT)           9357 non-null float64
PT08.S1(CO)      9357 non-null float64
NMHC(GT)         9357 non-null float64
C6H6(GT)         9357 non-null float64
PT08.S2(NMHC)    9357 non-null float64
NOx(GT)          9357 non-null float64
PT08.S3(NOx)     9357 non-null float64
NO2(GT)          9357 non-null float64
PT08.S4(NO2)     9357 non-null float64
PT08.S5(O3)      9357 non-null float64
T                9357 non-null float64
RH               9357 non-null float64
AH               9357 non-null float64
dtypes: float64(13), object(2)
memory usage: 1.1+ MB


In [9]:
#make sure that rows [9357:9471] are empty
df[9357:9471]

Date Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  PT08.S2(NMHC)  \
9357  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9358  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9359  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9360  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9361  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9362  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9363  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9364  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9365  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9366  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9367  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9368  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9369  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9370  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9371  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9372  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9373  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9374  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9375  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9376  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9377  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9378  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9379  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9380  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9381  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9382  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9383  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9384  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9385  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9386  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
...   ...  ...     ...          ...       ...       ...            ...   
9441  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9442  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9443  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9444  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9445  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9446  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9447  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9448  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9449  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9450  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9451  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9452  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9453  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9454  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9455  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9456  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9457  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9458  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9459  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9460  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9461  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9462  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9463  NaN

In [10]:
#drop empty rows 
df_dropna = df.dropna()
df_dropna.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9357 entries, 0 to 9356
Data columns (total 15 columns):
Date             9357 non-null object
Time             9357 non-null object
CO(GT)           9357 non-null float64
PT08.S1(CO)      9357 non-null float64
NMHC(GT)         9357 non-null float64
C6H6(GT)         9357 non-null float64
PT08.S2(NMHC)    9357 non-null float64
NOx(GT)          9357 non-null float64
PT08.S3(NOx)     9357 non-null float64
NO2(GT)          9357 non-null float64
PT08.S4(NO2)     9357 non-null float64
PT08.S5(O3)      9357 non-null float64
T                9357 non-null float64
RH               9357 non-null float64
AH               9357 non-null float64
dtypes: float64(13), object(2)
memory usage: 1.1+ MB


The df_dropna dataframe now contains 9357 non-null observations of 15 variables. However, according to the dataset description, missing values were recorded as '-200'. It is likely that malfunctioning sensors are to blame for these missing measurements.  These values, while not technically NaN, would still skew the results of the analysis. So, it is better to find and address them.

In [11]:
#replace -200.0 with NaN
df_withnulls = df_dropna.replace(-200.0, np.NaN)

In [12]:
#How many nulls were introduced?
df_withnulls.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9357 entries, 0 to 9356
Data columns (total 15 columns):
Date             9357 non-null object
Time             9357 non-null object
CO(GT)           7674 non-null float64
PT08.S1(CO)      8991 non-null float64
NMHC(GT)         914 non-null float64
C6H6(GT)         8991 non-null float64
PT08.S2(NMHC)    8991 non-null float64
NOx(GT)          7718 non-null float64
PT08.S3(NOx)     8991 non-null float64
NO2(GT)          7715 non-null float64
PT08.S4(NO2)     8991 non-null float64
PT08.S5(O3)      8991 non-null float64
T                8991 non-null float64
RH               8991 non-null float64
AH               8991 non-null float64
dtypes: float64(13), object(2)
memory usage: 1.1+ MB


As the output above shows, the NMHC(GT) column contains only 914 non-null values, 8623 (or slightly over 90%) of values are missing. Imputing these values might introduce bias in the dataset. So, in this case, it might be better just to drop this column.

In [13]:
#drop NMHC(GT) column
df_withnulls.drop('NMHC(GT)', axis=1, inplace=True)

In [14]:
#Check for remaining null values
df_withnulls.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9357 entries, 0 to 9356
Data columns (total 14 columns):
Date             9357 non-null object
Time             9357 non-null object
CO(GT)           7674 non-null float64
PT08.S1(CO)      8991 non-null float64
C6H6(GT)         8991 non-null float64
PT08.S2(NMHC)    8991 non-null float64
NOx(GT)          7718 non-null float64
PT08.S3(NOx)     8991 non-null float64
NO2(GT)          7715 non-null float64
PT08.S4(NO2)     8991 non-null float64
PT08.S5(O3)      8991 non-null float64
T                8991 non-null float64
RH               8991 non-null float64
AH               8991 non-null float64
dtypes: float64(12), object(2)
memory usage: 1.1+ MB


All remaining columns, except for the "Date" and "Time" have missing values. Since these measurements represent time series data from atmospheric sensors, it is likely that each data point is related to the points before and after.  So, we can use a variety of imputation methods - using a mean (median or most frequent number) for the column, as well as filling forward(backward) and interpolating. For practice purposes, I will use different methods for different columns.

In [15]:
#Fill forward  CO(GT) column - replace values with the last known value
df_withnulls.iloc[0:9356, 2].fillna(method='ffill', inplace=True)

In [16]:
#Check results
df_withnulls.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9357 entries, 0 to 9356
Data columns (total 14 columns):
Date             9357 non-null object
Time             9357 non-null object
CO(GT)           9357 non-null float64
PT08.S1(CO)      8991 non-null float64
C6H6(GT)         8991 non-null float64
PT08.S2(NMHC)    8991 non-null float64
NOx(GT)          7718 non-null float64
PT08.S3(NOx)     8991 non-null float64
NO2(GT)          7715 non-null float64
PT08.S4(NO2)     8991 non-null float64
PT08.S5(O3)      8991 non-null float64
T                8991 non-null float64
RH               8991 non-null float64
AH               8991 non-null float64
dtypes: float64(12), object(2)
memory usage: 1.1+ MB


In [18]:
#PT09.S1(CO) column - fill backwards - uses the newest values to replace a missing value
df_withnulls.iloc[0:9356, 3].fillna(method='bfill', inplace=True)

In [19]:
#Use interpolation to fill missing values - 'C6H6(GT)' column
#method "pad" - fill in NaNs using existing values
df_withnulls.iloc[:,4].interpolate(method='pad', inplace=True)

In [20]:
#Check results
df_withnulls.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9357 entries, 0 to 9356
Data columns (total 14 columns):
Date             9357 non-null object
Time             9357 non-null object
CO(GT)           9357 non-null float64
PT08.S1(CO)      9357 non-null float64
C6H6(GT)         9357 non-null float64
PT08.S2(NMHC)    8991 non-null float64
NOx(GT)          7718 non-null float64
PT08.S3(NOx)     8991 non-null float64
NO2(GT)          7715 non-null float64
PT08.S4(NO2)     8991 non-null float64
PT08.S5(O3)      8991 non-null float64
T                8991 non-null float64
RH               8991 non-null float64
AH               8991 non-null float64
dtypes: float64(12), object(2)
memory usage: 1.1+ MB


In [21]:
#copy remaining columns 
df_toimpute=df_withnulls.iloc[:, 5:14].values

In [23]:
df_toimpute.shape

(9357, 9)

In [25]:
#Use SimpleImputer from sklearn 
from sklearn.impute import SimpleImputer
#create an imputer using a mean value for a column
imputer=SimpleImputer(missing_values=np.nan, strategy='mean')
#fit it to the data
imputer.fit(df_toimpute)
#transform the data and convert it back to a panda data frame
df_nonulls =pd.DataFrame(imputer.transform(df_toimpute))

In [26]:
#check results
df_nonulls.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9357 entries, 0 to 9356
Data columns (total 9 columns):
0    9357 non-null float64
1    9357 non-null float64
2    9357 non-null float64
3    9357 non-null float64
4    9357 non-null float64
5    9357 non-null float64
6    9357 non-null float64
7    9357 non-null float64
8    9357 non-null float64
dtypes: float64(9)
memory usage: 658.0 KB


In [27]:
#add column names to the transformed data frame
df_nonulls.columns = ['PT08.S2(NMHC)', 'NOx(GT)', 'PT08.S3(NOx)', 'NO2(GT)', 'PT08.S4(NO2)', 'PT08.S5(O3)', 'T', 'RH', 'AH' ]

In [28]:
#Copy initial Date, Time columns and columns to which we applied filling forward, backward and interpolation
df_final = df_withnulls[['Date', 'Time', 'CO(GT)', 'PT08.S1(CO)', 'C6H6(GT)']]

In [29]:
#concatenate the two data franes cleaned of null values in one dataframe
df_result=pd.concat([df_final, df_nonulls], axis=1)

In [30]:
#chech the first five rows
df_result.head()

Date      Time  CO(GT)  PT08.S1(CO)  C6H6(GT)  PT08.S2(NMHC)  \
0  10/03/2004  18.00.00     2.6       1360.0      11.9         1046.0   
1  10/03/2004  19.00.00     2.0       1292.0       9.4          955.0   
2  10/03/2004  20.00.00     2.2       1402.0       9.0          939.0   
3  10/03/2004  21.00.00     2.2       1376.0       9.2          948.0   
4  10/03/2004  22.00.00     1.6       1272.0       6.5          836.0   

   NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)     T    RH  \
0    166.0        1056.0    113.0        1692.0       1268.0  13.6  48.9   
1    103.0        1174.0     92.0        1559.0        972.0  13.3  47.7   
2    131.0        1140.0    114.0        1555.0       1074.0  11.9  54.0   
3    172.0        1092.0    122.0        1584.0       1203.0  11.0  60.0   
4    131.0        1205.0    116.0        1490.0       1110.0  11.2  59.6   

       AH  
0  0.7578  
1  0.7255  
2  0.7502  
3  0.7867  
4  0.7888

In [31]:
#check the last five rows
df_result.tail()

Date      Time  CO(GT)  PT08.S1(CO)  C6H6(GT)  PT08.S2(NMHC)  \
9352  04/04/2005  10.00.00     3.1       1314.0      13.5         1101.0   
9353  04/04/2005  11.00.00     2.4       1163.0      11.4         1027.0   
9354  04/04/2005  12.00.00     2.4       1142.0      12.4         1063.0   
9355  04/04/2005  13.00.00     2.1       1003.0       9.5          961.0   
9356  04/04/2005  14.00.00     2.2       1071.0      11.9         1047.0   

      NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)     T    RH  \
9352    472.0         539.0    190.0        1374.0       1729.0  21.9  29.3   
9353    353.0         604.0    179.0        1264.0       1269.0  24.3  23.7   
9354    293.0         603.0    175.0        1241.0       1092.0  26.9  18.3   
9355    235.0         702.0    156.0        1041.0        770.0  28.3  13.5   
9356    265.0         654.0    168.0        1129.0        816.0  28.5  13.1   

          AH  
9352  0.7568  
9353  0.7119  
9354  0.6406  
9355  0.5139  
9356  0.5028

In [33]:
#Display detailed information about the structure of the data frame
#to make sure that there are no null values
df_result.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9357 entries, 0 to 9356
Data columns (total 14 columns):
Date             9357 non-null object
Time             9357 non-null object
CO(GT)           9357 non-null float64
PT08.S1(CO)      9357 non-null float64
C6H6(GT)         9357 non-null float64
PT08.S2(NMHC)    9357 non-null float64
NOx(GT)          9357 non-null float64
PT08.S3(NOx)     9357 non-null float64
NO2(GT)          9357 non-null float64
PT08.S4(NO2)     9357 non-null float64
PT08.S5(O3)      9357 non-null float64
T                9357 non-null float64
RH               9357 non-null float64
AH               9357 non-null float64
dtypes: float64(12), object(2)
memory usage: 1.1+ MB


In [34]:
#Display summary statistics for the variables (columns)
#No '-200.0' or other default values resent
df_result.describe()

CO(GT)  PT08.S1(CO)     C6H6(GT)  PT08.S2(NMHC)      NOx(GT)  \
count  9357.000000  9357.000000  9357.000000    9357.000000  9357.000000   
mean      2.082195  1103.389120    10.190392     939.153376   246.896735   
std       1.469801   219.934242     7.565771     261.560236   193.426632   
min       0.100000   647.000000     0.100000     383.000000     2.000000   
25%       1.000000   937.000000     4.400000     743.000000   112.000000   
50%       1.700000  1069.000000     8.300000     923.000000   229.000000   
75%       2.800000  1242.000000    14.000000    1105.000000   284.000000   
max      11.900000  2040.000000    63.700000    2214.000000  1479.000000   

       PT08.S3(NOx)      NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)            T  \
count   9357.000000  9357.000000   9357.000000  9357.000000  9357.000000   
mean     835.493605   113.091251   1456.264598  1022.906128    18.317829   
std      251.743954    43.920954    339.367559   390.612324     8.657639   
min      322.000000     2.000000    551.000000   221.000000    -1.900000   
25%      666.000000    86.000000   1242.000000   742.000000    12.000000   
50%      818.000000   113.091251   1456.264598   983.000000    18.300000   
75%      960.000000   133.000000   1662.000000  1255.000000    24.100000   
max     2683.000000   340.000000   2775.000000  2523.000000    44.600000   

                RH           AH  
count  9357.000000  9357.000000  
mean     49.234201     1.025530  
std      16.974801     0.395835  
min       9.200000     0.184700  
25%      36.600000     0.746100  
50%      49.234201     1.015400  
75%      61.900000     1.296200  
max      88.700000     2.231000